Create block filters before poll interval

rodion-lim-partior opened the pull request:
Fixes: #151 

Raising this PR to allow for block filters to be created on init instead of after the block polling interval.

rodion-lim-partior committed:
Create block filters before poll interval

Signed-off-by: rodion <rodion.lim@partior.com>

EnriqueL8 commented:
Thanks for the PR @rodion-lim-partior  - @Chengxuan could you take a look at this one?

Chengxuan reviewed (changes requested):
@rodion-lim-partior Thanks for spotting https://github.com/hyperledger/firefly-evmconnect/issues/151 !

The solution proposed is likely to make the problem happen less frequently as long as no transactions are submitted before the new block filter is created on the node.

However, there is still a race condition. I think a better fix here is to ensure the `checkStartedLocked` function doesn't return until a block filter is established.

Please also add a test to catch future regressions.

rodion-lim-partior commented:
@Chengxuan, noticed the following when I tried to fix the race condition:

1. When we add a consumer to the block listener, we acquire a lock
2. Call `checkStartedLocked`, which starts the block listener loop in a separate goroutine
3. Establish block height next, which requires acquiring a lock to write to block listener, this is blocked till `checkStartedLocked` has finished
4. We then create a block filter. If I block `checkStartedLocked` until the filter is created, that ends up in a deadlock, since `checkStartedLocked` is required to finish to release the lock before establishing block height can proceed

Chengxuan commented:
@rodion-lim-partior It feels the problem is within the scope of `checkStartedLocked` function, so the solution shouldn't need to touch the global lock that is used to protect shared resources for concurrency. 

The key problem to solve is synchronization between the  `block listener loop` separate go routine and the main go routine that created that separate go routine.

Using a channel can also be a solution and feels more idiomatic in Golang.

rodion-lim-partior committed:
Block listener should notify when block filter is established

Signed-off-by: rodion <rodion.lim@partior.com>

rodion-lim-partior commented:
@Chengxuan, can you help to take a look at the current commit if this was what you referred to?

Chengxuan commented on internal/ethereum/blocklistener.go:
```diff
@@ -185,6 +194,8 @@ func (bl *blockListener) listenLoop() {
 				failCount++
 				continue
+			} else {
+				close(bl.blockFilterEstablished)
```
need a check here to make sure closing the channel is only attempted once.

rodion-lim-partior replied:
done now

Chengxuan commented:
@rodion-lim-partior Changes look great! I pointed out one possible issue in the code, you should be able to produce that error case in a test.

Once tests are added, I'm happy to take another look. 

Thank you for the work!

EnriqueL8 commented:
Thank @Chengxuan and @rodion-lim-partior for the collaboration on this one! Looking forward to seeing some tests

rodionlim commented:
@Chengxuan, the commit I pushed causes a deadlock in my local environment due to the reasons in my previous comments. I will have to shift the global locks downwards to resolve the deadlocks. From local testing, shifting the locks down does not seem to break the existing implementation. Let me know if that works, I will implement the tests and refactor based on your comments after.

Chengxuan commented:
Hi @rodion-lim-partior I think you have a good understanding of the code now. So please feel free to try out different options and write tests. Please make sure the logic must ensure only 1 listener loop go routine can be created after any change. Feel free to split global mutex to be resource specific if needed.

rodionlim commented:
@Chengxuan, got it. Will do some refactoring of the code and come back with the tests for another round of review. Thanks for the pointers!

rodion-lim-partior committed:
Correct failing tests

Signed-off-by: rodion <rodion.lim@partior.com>

rodion-lim-partior committed:
Add tests for defering checkStartedLocked till block filters are created

Signed-off-by: rodion <rodion.lim@partior.com>

rodion-lim-partior commented:
@Chengxuan @EnriqueL8, apologies for the delay. I refactored the code a fair bit and fixed some of the race conditions along the way. Added the tests as well for the edge case and the happy case scenario.

Some observations (correct me if I'm wrong):
1. There are two functions `addConsumer` and `getHighestBlock` which used to lock `checkStartedLocked` itself, leading to behavior which is difficult to reason about and unintentionally cause certain logic to occur in a particular sequence (described in a previous comment)
2. There is a small chance that a consumer is not added to the block listener (because the thread is slow etc.), causing mock `getFilterChanges` calls in the tests to skip the dispatch call, resulting in failed tests. This has now been fixed as well
3. Now that we wait for block filters to be established, many of the tests in event_actions.go and event_listener.go need to mock the RPC calls to create the block filters and listen on changes, this has also been corrected.

EnriqueL8 commented on internal/ethereum/blocklistener.go:
```diff
@@ -470,23 +518,28 @@ func (bl *blockListener) dispatchToConsumers(consumers []*blockUpdateConsumer, u
 		go bl.listenLoop()
 	}
+
+	bl.blockTillBlockFilterEstablished(ctx)
```
why would we do this for every consumer? If the `listenLoop` is already running before. I think we only need this once

EnriqueL8 replied:
This would block adding new consumers

rodion-lim-partior replied:
within bl.blocktillblockfilterestablished, it checks if the channel is closed, if the loop is already running, the channel will be closed, this won't block

EnriqueL8 replied:
We do if we put it inside the if block

rodionlim replied:
Makes sense, I'll test it and shift it inside the if block if nothing breaks

Chengxuan replied:
what @rodion-lim-partior has done here makes sense to me. Using a mutex to ensure that there will be only 1 instance of `listenLoop` to close the `signal` channel while ensuring all other go routines wait until the `signal` channel is closed matches the behavior we discussed.

EnriqueL8 replied:
Okay, so when I wrote this comment it was in conjunction with the call to `checkStartedLocked` still being locked by `bl.mux.Lock()` but yes if we are now removing it we would need a lock at this point

EnriqueL8 commented on internal/ethereum/blocklistener.go:
```diff
@@ -470,23 +518,28 @@ func (bl *blockListener) dispatchToConsumers(consumers []*blockUpdateConsumer, u
 	bl.mux.Lock()
-	defer bl.mux.Unlock()
-	bl.checkStartedLocked()
 	bl.consumers[*c.id] = c
+	bl.mux.Unlock()
+	bl.checkStartedLocked(context.Background())
```
We would propagate the correct context here

EnriqueL8 replied:
and we need to keep it in the lock because we check the state of the block listener in checkStartedLocked with `bl.listenLoopDone == nil ` There could be a race condition where two loops get started if not

rodion-lim-partior replied:
@EnriqueL8, added a patch to lock till the listenLoopDone is set

EnriqueL8 commented on internal/ethereum/blocklistener.go:
```diff
@@ -92,6 +100,34 @@ func newBlockListener(ctx context.Context, c *ethConnector, conf config.Section,
+func (bl *blockListener) blockTillBlockFilterEstablished(ctx context.Context) {
+	select {
+	case <-bl.blockFilterEstablished.signal:
+	case <-bl.ctx.Done():
+	case <-ctx.Done():
```
not sure why we need both here

rodion-lim-partior replied:
context.Done is for client side timeout by functions like `listener.getInitialBlock`, `bl.ctx.Done` takes care of cases like a sigint

EnriqueL8 commented on internal/ethereum/blocklistener.go:
```diff
@@ -185,6 +231,8 @@ func (bl *blockListener) listenLoop() {
 				log.L(bl.ctx).Errorf("Failed to establish new block filter: %s", err.Message)
 				failCount++
 				continue
+			} else {
```
nit: There is a continue above so no else needed

rodion-lim-partior replied:
good catch, thanks

EnriqueL8 reviewed (commented):
Looking good - a few thoughts

rodion-lim-partior committed:
Lock block listener loop started flag

Signed-off-by: rodion <rodion.lim@partior.com>

Chengxuan commented on internal/ethereum/blocklistener.go:
```diff
@@ -92,6 +100,34 @@ func newBlockListener(ctx context.Context, c *ethConnector, conf config.Section,
+	}
+}
+
+func (bl *blockListener) blockTillBlockFilterEstablished(ctx context.Context) {
```
I suggest removing `ctx context.Context` from this function as it's not used in the way you described. @rodion-lim-partior To enable it to work in the way you described, you'll need to surface this extra parameter to the top level. For now, in the code, you just used `context.Background()` which only gets canceled when the program exit, seems a bit redundant to `bl.ctx.Done()`.

Chengxuan commented on internal/ethereum/blocklistener.go:
```diff
@@ -136,22 +172,27 @@ func (bl *blockListener) establishBlockHeightWithRetry() error {
 
 func (bl *blockListener) listenLoop() {
 	defer close(bl.listenLoopDone)
+	defer func() {
+		bl.unsetBlockFilterStatus()
+	}()
```
should not need this function unless you'd like to introduce the behavior of un-initialising block listener which needs extra design itself.

Chengxuan commented on internal/ethereum/blocklistener.go:
```diff
@@ -40,6 +40,12 @@ type blockUpdateConsumer struct {
+type blockFilterStatus struct {
+	isEstablished bool
+	signal        chan struct{}
+	mux           sync.Mutex
```
don't think this mutex is actually required as the upstream logic already ensures there is a single thread modify the channel, so a flag + a channel should does the job.

Can we name them consistently with existing methods name. e.g. `isStarted` & `startDone`?

Chengxuan commented:
@rodionlim logic looks good to me 👍 , some minor comments to clean up bits that are not necessary.

rodion-lim-partior committed:
Fix PR review comments

Signed-off-by: rodion <rodion.lim@partior.com>

rodion-lim-partior commented:
@Chengxuan, pushed a commit to clean up the code, left with one pending comment that needs a review

Chengxuan commented:
@rodion-lim-partior thanks. For `one pending comment that needs a review` I assume that's referring to the removal of ctx parameter in `blockTillBlockFilterEstablished` that you are reviewing and addressing. 

If there are something you'd like me to review, please let me know.

rodion-lim-partior commented:
@EnriqueL8, yep that's right. Rest of your review comments have been addressed. The pending one that needs a review, I need @Chengxuan to take a look. Removing the context will cause test failures.

Chengxuan committed:
refactor tests

Signed-off-by: Chengxuan Xing <chengxuan.xing@kaleido.io>

Chengxuan commented:
@rodion-lim-partior I looked at the code closer and understood the need to have the context, here is the proposed change: https://github.com/partior-3p/firefly-evmconnect/pull/1

Chengxuan committed:
downscoping the locking for get highest block

Signed-off-by: Chengxuan Xing <chengxuan.xing@kaleido.io>

Chengxuan committed:
more tidy up

Signed-off-by: Chengxuan Xing <chengxuan.xing@kaleido.io>

Chengxuan committed:
fixing test coverage

Signed-off-by: Chengxuan Xing <chengxuan.xing@kaleido.io>

Chengxuan committed:
Merge branch 'main' of github.com:kaleido-io/firefly-evmconnect into fix-long-poll

Chengxuan committed:
fixing tests

Signed-off-by: Chengxuan Xing <chengxuan.xing@kaleido.io>

Chengxuan commented:
Hi @rodion-lim-partior I'm planning to do a release to pick up another bug fix tomorrow. I'd like to get this PR in as well, please take a look at https://github.com/partior-3p/firefly-evmconnect/pull/1 and merge it if you are happy with it. I've sorted out the merge conflicts in that PR as well

rodion-lim-partior committed:
Fix lint failures

Signed-off-by: rodion <rodion.lim@partior.com>

rodion-lim-partior commented:
@Chengxuan, merged those changes in. Corrected some of the lint failures as well when you run `make test`. Seems to be working fine on my local environment

Chengxuan reviewed (approved):
Awesome. Thanks for your contribution @rodion-lim-partior 👍

rodion-lim-partior commented:
Cheer guys.
